app/utils.py

```diff
@@ -68,18 +85,39 @@ async def _make_request(request,
                      method=method,
                      url=url,
                      handler=func.__name__)
+
+        attempt_number = View._make_request.retry.statistics['attempt_number']
         try:
-            async with request.app.http_session_pool.request(
-                    method, url, auth=auth, json=json, ssl=False) as resp:
-                func(resp)
-                if return_json:
-                    return await resp.json()
-                else:
-                    return None
+            if attempt_number > 1:
+                wait_secs = int(request.app['WAIT_BEFORE_RETRY'])
+                time.sleep(wait_secs)
+                # basic request without keep-alive to avoid terminating service.
+                logger.info('retrying using basic connection', attempt_number=attempt_number, wait_secs=wait_secs)
+                async with aiohttp.request(
+                        method, url, auth=auth, json=json) as resp:
+                    func(resp, attempt_number)
+                    if return_json:
+                        return await resp.json()
+                    else:
+                        return None
+            else:
+                # normal path. pooled ; keep-alive request for performance
+                async with request.app.http_session_pool.request(
+                        method, url, auth=auth, json=json, ssl=False) as resp:
+                    func(resp, attempt_number)
+                    if return_json:
+                        return await resp.json()
+                    else:
+                        return None
         except (ClientConnectionError, ClientConnectorError) as ex:
-            logger.error('client failed to connect',
-                         url=url,
-                         client_ip=request['client_ip'])
+            if attempt_number < attempts_retry_limit:
+                logger.warn('client failed to connect, could be during service scale back',
```

Comment:
Just wondering but is it the case that when a RHSVC instance exits that you get a bunch of these warnings, and then after after about 5 seconds it'll get a new connection and the warnings stop?

Reply:
As we discussed, I think the sequence is this:

1. scale-back / pod delete k8s instruction issued . 
2. pod taken out of service (clusterIP / loadbalancer) IP tables, BUT RHUI-RHSvc connections continue , even it seems beyond any keep-alive timeout
3. prestop of 5 seconds, during which all connection are happy (because nothing has closed yet)
4. SIGTERM issued from k8s to RHSvc
5. Undertow graceful shutdown (usually less than a second) . Start of 503's back to RHUI
6. RHSvc application-context shutsdown
7. RHSvc pod terminates . A smattering of socket connection exceptions back at RHUI, followed by everything working fine with no exceptions.

During 503's and connection exceptions, the code I have written will retry, rerouting to another live RHSvc pod, to complete normally.

I think about 10-15 connections are affected during a pod shutdown at load when there is about 6,000 users.